File: code_samples/section13/example_2_prefix_counting/trie_prefix_count.java
package code_samples.section13.example_2_prefix_counting;

import java.io.BufferedReader; // Efficient line-by-line reading of text files
import java.io.FileReader;     // Character stream for reading a file
import java.io.IOException;    // Exception type for I/O failures

/*
 * Trie Prefix Counting (Java)
 *
 * This example demonstrates a trie (prefix tree) that supports:
 * - Inserting words (case-insensitive)
 * - Counting how many inserted words start with a given prefix
 *
 * Key constraints:
 * - Only letters 'a'–'z' are accepted (case-insensitive).
 * - If any character in a word is not a–z, insert() rejects the entire word.
 * - prefixCount(prefix) returns 0 if the prefix is invalid or not found.
 *
 * Dictionary loading:
 * - Reads words from a file (default: "..\\data\\words.txt")
 * - Relative paths resolve from the working directory where Java is executed.
 */
public class trie_prefix_count {

    /*
     * Number of supported lowercase English letters.
     */
    private static final int ALPHABET_SIZE = 26;

    /*
     * Trie node definition.
     *
     * Each node contains:
     * - children: references to child nodes for each letter 'a'–'z'
     * - prefixCount: number of inserted words that share this prefix node
     * - isEndOfWord: end-of-word marker (not used by prefix counting output,
     *   but set during insert for completeness / consistency with other examples)
     */
    private static class TrieNode {
        TrieNode[] children = new TrieNode[ALPHABET_SIZE];

        // Number of words whose insertion path includes this node
        int prefixCount = 0;      // number of words sharing this prefix

        // Included for compatibility with word-search examples; not used here
        @SuppressWarnings("unused")
        boolean isEndOfWord = false;
    }

    /*
     * Root node of the trie.
     * Always exists and does not correspond to any particular character.
     */
    private final TrieNode root = new TrieNode();

    /*
     * Convert a lowercase character to its child index.
     *
     * Parameters:
     * - c: character expected to be between 'a' and 'z'
     *
     * Returns:
     * - 0..25 if valid
     * - -1 if invalid
     */
    private static int index(char c) {
        if (c < 'a' || c > 'z') return -1;
        return c - 'a';
    }

    /**
     * Inserts a word into the trie and increments prefixCount along the path.
     *
     * Behavior:
     * - Normalizes characters to lowercase (case-insensitive)
     * - Rejects the entire word if any character is not 'a'–'z'
     * - Creates nodes lazily as needed
     * - Increments prefixCount after advancing into the child node, meaning
     *   the node's prefixCount represents how many words share that prefix node
     *
     * @param word Word to insert into the trie.
     */
    public void insert(String word) {
        TrieNode current = root;

        // Traverse each character in the word
        for (int i = 0; i < word.length(); i++) {
            char c = Character.toLowerCase(word.charAt(i));
            int idx = index(c);

            // Reject entire word if invalid character is encountered
            if (idx < 0) return; // reject word

            // Create the child node if this path does not yet exist
            if (current.children[idx] == null) {
                current.children[idx] = new TrieNode();
            }

            // Move to the next node in the trie
            current = current.children[idx];

            // Increment prefixCount for this prefix node
            current.prefixCount++;
        }

        // Mark end-of-word (not used by prefixCount(), but kept for completeness)
        current.isEndOfWord = true;
    }

    /**
     * Returns how many words in the trie start with the given prefix.
     *
     * Behavior:
     * - Returns 0 if the prefix contains invalid characters (non a–z)
     * - Returns 0 if the prefix path does not exist in the trie
     * - Otherwise returns prefixCount stored in the node representing the prefix
     *
     * @param prefix Prefix to query.
     * @return Count of inserted words that start with the prefix.
     */
    public int prefixCount(String prefix) {
        TrieNode current = root;

        // Walk the trie according to the prefix characters
        for (int i = 0; i < prefix.length(); i++) {
            char c = Character.toLowerCase(prefix.charAt(i));
            int idx = index(c);

            // Invalid character -> prefix is invalid by design
            if (idx < 0) return 0;

            // Missing child -> prefix does not exist in trie
            if (current.children[idx] == null) return 0;

            // Advance to the next node
            current = current.children[idx];
        }

        // The node's prefixCount stores how many words share this prefix
        return current.prefixCount;
    }

    /*
     * Loads words from a dictionary file into the trie.
     *
     * Parameters:
     * - trie: trie instance to populate
     * - path: filesystem path to dictionary file
     *
     * Returns:
     * - number of non-empty lines processed (insert attempted for each)
     * - 0 if the file cannot be opened/read
     *
     * Notes:
     * - insert() may reject words containing invalid characters, but the loader
     *   still increments the count for each non-empty trimmed line (as written).
     */
    private static int loadDictionary(trie_prefix_count trie, String path) {
        int count = 0;

        // Try-with-resources ensures the reader is closed even on error
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            String line;

            // Read the file line by line
            while ((line = br.readLine()) != null) {
                line = line.trim();

                // Skip empty lines
                if (line.isEmpty()) continue;

                // Insert word into the trie (may be rejected if invalid)
                trie.insert(line);
                count++;
            }
        } catch (IOException e) {
            // Report file access / I/O errors
            System.err.println("Failed to open dictionary: " + path);
            System.err.println(e.getMessage());
            return 0;
        }

        return count;
    }

    /*
     * Program entry point.
     *
     * Demonstrates:
     * - Dictionary loading
     * - Prefix-count queries
     *
     * Arguments:
     * - args[0] (optional): dictionary path; defaults to "..\\data\\words.txt"
     */
    public static void main(String[] args) {
        trie_prefix_count trie = new trie_prefix_count();

        // Default path matches lesson convention; allow override via CLI arg
        String dictPath = (args.length > 0) ? args[0] : "..\\data\\words.txt";

        // Load dictionary words into the trie
        int loaded = loadDictionary(trie, dictPath);

        System.out.println("Loaded " + loaded + " words from " + dictPath);
        System.out.println();

        // Prefixes to query
        String[] prefixes = { "a", "ab", "alg", "aard", "z", "nope" };
        for (String p : prefixes) {
            System.out.println("prefixCount(\"" + p + "\") = " + trie.prefixCount(p));
        }
    }
}

File: code_samples/section13/example_2_prefix_counting/trie_prefix_count.cs
// Trie Prefix Counting (top-level, dotnet-script friendly)
//
// This script implements a trie (prefix tree) focused on prefix counting:
// given a prefix, return how many inserted words start with that prefix.
//
// Key behavior / constraints:
// - Only letters 'a'–'z' are accepted (case-insensitive).
// - If any character in a word is not a–z, Insert() rejects the entire word.
// - PrefixCount(prefix) returns 0 if the prefix is invalid or not present.
// - Dictionary loading reads words from a file (default: ..\data\words.txt).
// - Relative paths are resolved from the working directory where the script runs.

const int ALPHABET_SIZE = 26; // Number of supported lowercase letters ('a'–'z')

// Root node of the trie; acts as the entry point for all insert/query operations
TrieNode root = new();

/*
 * Convert a lowercase character into a child-array index.
 *
 * Parameters:
 * - c: expected to be 'a'–'z'
 *
 * Returns:
 * - 0..25 for valid 'a'–'z'
 * - -1 for any invalid character
 */
int Index(char c) => (c >= 'a' && c <= 'z') ? (c - 'a') : -1;

/*
 * Insert a word into the trie.
 *
 * Behavior:
 * - Converts each character to lowercase (case-insensitive handling)
 * - Rejects the entire word if any character is not 'a'–'z'
 * - Creates nodes lazily as needed
 * - Increments PrefixCount on each node visited along the path
 *
 * Note:
 * - PrefixCount is incremented after advancing into the child node, meaning
 *   it represents "how many inserted words share this prefix node".
 */
void Insert(string word)
{
    TrieNode current = root;

    // Traverse the trie character-by-character
    foreach (char raw in word)
    {
        // Normalize character to lowercase using invariant rules
        char c = char.ToLowerInvariant(raw);

        // Map character to a 0..25 index
        int idx = Index(c);

        // Reject entire word if an invalid character is encountered
        if (idx < 0) return; // reject word

        // Allocate a child node if it doesn't exist
        if (current.Children[idx] == null)
            current.Children[idx] = new TrieNode();

        // Move to the next node in the trie
        current = current.Children[idx];

        // Increment count of words that share this prefix (i.e., this node)
        current.PrefixCount++;
    }
}

/*
 * Return how many inserted words start with the given prefix.
 *
 * Behavior:
 * - Returns 0 if the prefix contains invalid characters (non a–z)
 * - Returns 0 if the prefix path does not exist in the trie
 * - Otherwise returns PrefixCount stored in the node representing the prefix
 *
 * Parameters:
 * - prefix: prefix string to query
 *
 * Returns:
 * - number of words sharing that prefix (0 if invalid / not found)
 */
int PrefixCount(string prefix)
{
    TrieNode current = root;

    // Walk the trie according to the prefix
    foreach (char raw in prefix)
    {
        char c = char.ToLowerInvariant(raw);
        int idx = Index(c);

        // Invalid character -> prefix cannot exist by design
        if (idx < 0) return 0;

        // Missing path -> no words share this prefix
        if (current.Children[idx] == null) return 0;

        current = current.Children[idx];
    }

    // The stored PrefixCount represents how many words share this prefix node
    return current.PrefixCount;
}

/*
 * Load a dictionary file (one word per line) into the trie.
 *
 * Parameters:
 * - path: filesystem path to the dictionary file
 *
 * Returns:
 * - number of non-empty lines processed (Insert() attempted for each)
 * - 0 if the file cannot be opened or read
 *
 * Notes:
 * - Insert() may reject words containing invalid characters, but this loader
 *   still increments "count" for every non-empty trimmed line (as written).
 */
int LoadDictionary(string path)
{
    int count = 0;

    try
    {
        // File.ReadLines streams lines lazily, suitable for large files
        foreach (var line in File.ReadLines(path))
        {
            // Trim whitespace and skip empty lines
            var w = line.Trim();
            if (w.Length == 0) continue;

            // Insert the word into the trie (may be rejected if invalid)
            Insert(w);
            count++;
        }
    }
    catch (Exception ex)
    {
        // Report file access / I/O errors
        Console.Error.WriteLine($"Failed to open dictionary: {path}");
        Console.Error.WriteLine(ex.Message);
        return 0;
    }

    return count;
}

// -------------------- Main --------------------

/*
 * Resolve dictionary path.
 * If a script argument is provided, use it; otherwise use the lesson default.
 */
string dictPath = (Args.Count > 0) ? Args[0] : @"..\data\words.txt";

// Load dictionary words into the trie
int loaded = LoadDictionary(dictPath);

Console.WriteLine($"Loaded {loaded} words from {dictPath}\n");

// Prefixes to query against the loaded trie
string[] prefixes = ["a", "ab", "alg", "aard", "z", "nope"];
foreach (var p in prefixes)
    Console.WriteLine($"prefixCount(\"{p}\") = {PrefixCount(p)}");

Console.WriteLine();

// Helpful diagnostics for debugging relative paths at runtime
Console.WriteLine($"Working directory = {Directory.GetCurrentDirectory()}");
Console.WriteLine($"Resolved dictionary path = {Path.GetFullPath(dictPath)}");

/*
 * Trie node definition.
 *
 * Each node contains:
 * - Children: 26 child references (for 'a'–'z')
 * - PrefixCount: number of inserted words that include this node in their path
 *
 * This script focuses on prefix counting, so there is no end-of-word flag here.
 */
class TrieNode
{
    public TrieNode[] Children = new TrieNode[ALPHABET_SIZE]; // Child nodes
    public int PrefixCount = 0; // number of words sharing this prefix
}
